File: 12/one.lua
-- --- Advent of code 2021: Day 12 ---

-- (Skeleton file automatically created by aocTool, developed by Bolu, 2020-2021.)

package.path = package.path .. ";../modules/?.lua"
local tools = require "tools"

caves = tools.readlines("input.txt")

connections = {}
small_caves = {}
for _, cave in pairs(caves) do
	org, dst = table.unpack(tools.split(cave, "-"))
	if not connections[org] then
		connections[org] = {}
	end
	connections[org][dst] = true
	if dst ~= "end" and org ~= "start" then
		if not connections[dst] then
			connections[dst] = {}
		end
		connections[dst][org] = true
	end
	if org:sub(1, 1) >= 'a' then
		small_caves[org] = true
	end
end
small_caves["end"] = nil
small_caves.start = nil

function explore(node, caves, small_caves, explored, path, paths)
	for nxt, _ in pairs(caves[node]) do
		if nxt == "end" then
			paths[#paths + 1] = path .. ",end"
		elseif not explored[nxt] then
			explored_nxt = {}
			for cave, _ in pairs(explored) do
				explored_nxt[cave] = true
			end
			if small_caves[nxt] then
				explored_nxt[nxt] = true
			end
			explore(nxt, caves, small_caves, explored_nxt, path .. "," .. tostring(nxt), paths)
		end
	end
end

paths = {}
explore("start", connections, small_caves, {}, "start", paths)
print("Amount of paths that only visit small caves at least once: " .. #paths)
assert(#paths == 4104)

File: 11/one.py
# --- Advent of code 2021: Day 11 ---

# (Skeleton file automatically created by aocTool, developed by Bolu, 2020-2021.)

energy = [[int(y) for y in list(x)] for x in open("input.txt").read().splitlines()]

flashes = 0
for _ in range(100):
	# 1
	second_step = False
	for i in range(10):
		for j in range(10):
			energy[i][j] += 1
			if energy[i][j] > 9:
				second_step = True

	# 2
	while second_step:
		second_step = False
		for i in range(10):
			for j in range(10):
				if energy[i][j] > 9:
					flashes -=- 1
					for di in range(-1, 2):
						for dj in range(-1, 2):
							if i + di >= 0 and i + di < 10 and j + dj >= 0 and j + dj < 10:
								if di or dj:
									energy[i + di][j + dj] += 1
									if energy[i + di][j + dj] > 9 and (di < 0 or dj < 0):
										second_step = True
					energy[i][j] = -energy[i][j]

	# 3
	for i in range(10):
		for j in range(10):
			if energy[i][j] < 0:
				energy[i][j] = 0

print(f"Flashes: {flashes}")
assert flashes == 1617
